File: third/app/src/main/java/com/example/chapter3/homework/PlaceholderFragment.java
package com.example.chapter3.homework;

import android.animation.Animator;
import android.animation.AnimatorInflater;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import com.airbnb.lottie.LottieAnimationView;

import java.util.Objects;

public class PlaceholderFragment extends Fragment {


    LottieAnimationView loadingView;
    ListView firendListView;
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // TODO ex3-3: 修改 fragment_placeholder，添加 loading 控件和列表视图控件
        View view = inflater.inflate(R.layout.fragment_placeholder,container,false);
        String[] data = getArguments().getStringArray("data");
        loadingView = view.findViewById(R.id.loading_view);
        firendListView = new ListView(getContext());
        firendListView.setAdapter(new ArrayAdapter<>(Objects.requireNonNull(getContext()),android.R.layout.simple_expandable_list_item_1,data));
        firendListView.setVisibility(View.VISIBLE);
        firendListView.setAlpha(0f);
        ViewGroup viewGroup = (ViewGroup)view;
        viewGroup.addView(firendListView);
        return viewGroup;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        getView().postDelayed(new Runnable() {
            @Override
            public void run() {
                // 这里会在 5s 后执行
                // TODO ex3-4：实现动画，将 lottie 控件淡出，列表数据淡入
                Animator fadeOutAnimator = AnimatorInflater.loadAnimator(getContext(), R.animator.fade_out);
                final Animator fadeInAnimator = AnimatorInflater.loadAnimator(getContext(), R.animator.fade_in);
                fadeOutAnimator.setTarget(loadingView);
                fadeInAnimator.setTarget(firendListView);
                fadeOutAnimator.addListener(new Animator.AnimatorListener() {
                    @Override
                    public void onAnimationStart(Animator animator) { }

                    @Override
                    public void onAnimationEnd(Animator animator) {
                        fadeInAnimator.start();
                        ViewGroup container = (ViewGroup)loadingView.getParent();
                        container.removeView(loadingView);
                    }

                    @Override
                    public void onAnimationCancel(Animator animator) { }

                    @Override
                    public void onAnimationRepeat(Animator animator) { }
                });
                fadeOutAnimator.start();
            }
        }, 5000);
    }
}

File: third/app/src/main/java/com/example/chapter3/homework/PlaceholderFragment.kt
package com.example.chapter3.homework
import android.animation.Animator
import android.animation.AnimatorInflater
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ListView
import com.airbnb.lottie.LottieAnimationView

class PlaceholderFragment: Fragment(){
    private var loadingView: LottieAnimationView? = null
    private var friendListView: ListView? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        //用于载入res/layout下的xml布局文件并实例化，类似findById()
        val view = inflater.inflate(R.layout.fragment_placeholder,container,false)
        val data = arguments?.getStringArray("data")!!//如果为空，就抛出异常
        loadingView = view.findViewById(R.id.loading_view)
        friendListView = ListView(context)//返回当前view运行在那个activity context当中
        friendListView!!.adapter = ArrayAdapter(context!!,android.R.layout.simple_expandable_list_item_1,data)//android.R即为系统预先定义的，第三个args传入listStrings即可
        friendListView!!.visibility = View.VISIBLE
        friendListView!!.alpha = 0f
        (view as ViewGroup).addView(friendListView)
        return view
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        view!!.postDelayed({
            val fadeoutAnimationView = AnimatorInflater.loadAnimator(context,R.animator.fade_out)
            val fadeinAnimationView = AnimatorInflater.loadAnimator(context,R.animator.fade_in)
            fadeoutAnimationView.setTarget(loadingView)
            fadeinAnimationView.setTarget(friendListView)
            fadeoutAnimationView.addListener(object : Animator.AnimatorListener{
                override fun onAnimationCancel(animator: Animator?) {}
                override fun onAnimationEnd(animator: Animator?) {
                    fadeinAnimationView.start()
                    val container = (loadingView?.parent as ViewGroup)
                    container.removeView(loadingView)
                }
                override fun onAnimationStart(animator: Animator?) {}
                override fun onAnimationRepeat(animator: Animator?) {}
            })
            fadeoutAnimationView.start()
        },5000)

    }
}